flask_praetorian/base.py

```diff
@@ -484,3 +550,111 @@ def pack_header_for_user(
+                tmpl = Template(_template.read())
+            notification.message = tmpl.render(notification.__dict__).strip()
+            subject = subject if subject else self.confirmation_subject
+
+            msg = Message(body=notification.message,
+                          sender=self.confirmation_sender,
+                          subject=subject,
+                          recipients=[notification.email])
+
+            from flask import current_app as app
+            notification.errors = app.mail.send(msg)
+
+        return notification
+
+    def validate_confirmation(self, token=None):
+        return self.extract_jwt_token(token)
+
+    def hash_password(self, raw_password):
+        """
+        Encrypts a plaintext password using the stored passlib password context
+        """
+        PraetorianError.require_condition(
+            self.pwd_ctx is not None,
+            "Praetorian must be initialized before this method is available",
+        )
+        """
+        `scheme` is now set with self.pwd_ctx.update(default=scheme) due
+            to the depreciation in upcoming passlib 2.0.
+         zillions of warnings suck.
+        """
+        return self.pwd_ctx.hash(raw_password)
+
+    def verify_and_update(self, user=None, password=None):
+        """
+        Validate a password hash contained in the user object is in
+             hashed with the defined hash scheme (PRAETORIAN_HASH_SCHEME).
+        If not, raise an Exception of `InsecureHash`, unless the
+             `password` arguement is provided, in which case an attempt
+             to call `user.save()` will be made, updating the hashed
+             password to the currently desired hash scheme
+             (PRAETORIAN_HASH_SCHEME).
+
+        :param: user:     The user object to tie claim to
+                              (username, id, email, etc). *MUST*
+                              include the hashed password field,
+                              defined as `user.password`
+        :param: password: The user's provide password from login.
+                          If present, this is used to validate
+                              and then attempt to update with the
+                              new PRAETORIAN_HASH_SCHEME scheme.
+        """
+        if self.pwd_ctx.needs_update(user.password):
+            if password:
+                rv, updated = self.pwd_ctx.verify_and_update(password,
+                                                             user.password)
+                if rv:
+                    user.password = updated
+                else:
+                    raise AuthenticationError("Could not verify password")
+            else:
+                used_hash = self.pwd_ctx.identify(user.password)
+                desired_hash = self.hash_scheme
+                raise LegacyScheme("Hash using non-current scheme '{}'"
+                                   .format(used_hash, desired_hash))
```

Comment:
passing two format args, but only one `{}` in the string.

Reply:
submitted